### DISCUSSION TITLE
Locals of type 'ref struct' in async methods

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#4687] [Language Ideas]
My understanding is that, as long as the underlying storage is not `stakcalloc`, this should be safe.

```cs
async Task M() {
  Span<int> t = new int[] { 1, 2, 3 };
  ..
}
```

There are certain APIs that could use span for efficiency but that would currently limit the usage to non-async methods.

AFAIK there is no runtime restriction here.

**Comment by jnm2** [comment]
As long as the span is not accessed on the other side of an `await` which would force it onto the heap, I don't know of a reason why this shouldn't be allowed. It would make the async state machine generation code a little more complicated.

This limitation is a fairly common irritation for me. I get it when it has to survive an await, but often it doesn't.

**Reply by alrz** [reply]
Note that the motivation here *in async code* is not performance - so heap allocation of span is fine - it would rather enable a unified API surface for both sync and async code. In the former you must pass an array but the underlying impl only operates on span.

To give a little background, I'm exploring buffering strategies for enumerable list patterns. A span-backed type is frobbiden in async code so we may need to rely on two different APIs. The compiler *could* probably just emit such code, but there's no reason to not expose this feature to the language.

**Reply by bernd5** [reply]
`Span<T>` can not be stored on the heap - just the array itself.
Local `Span<T>` could be used theoretically without crossing await boundaries - but to simplify all that - `Span<T>` locals are prohibited in async methods.

But you can write:
```cs
using System;
using System.Threading.Tasks;

class App{

	int DoWithSpan(Span<int> i){
		return i[0];
	}
	async Task M() {
		var array = new int[] { 1, 2, 3 };
		DoWithSpan(array.AsSpan());
	}
}
```

**Reply by jnm2** [reply]
@alrz The runtime does not allow heap allocation of ref-like types (including Span) because this is unsafe. It's not a question of performance.

**Reply by alrz** [reply]
Couldn't find @stephentoub's comment on that. Regardless, a lot of discussions is happening on `ref struct InterpolatedSpanBuilder` I think this aspect will be somehow "resolved", one way or another.

**Reply by 333fred** [reply]
> Regardless, a lot of discussions is happening on ref struct InterpolatedSpanBuilder I think this aspect will be somehow "resolved", one way or another.

I wouldn't count on that. We can special case a particular pattern that is easily guaranteed to not be crossing async boundaries without generalizing it to _all_ ref structs used in these locations.

**Reply by acaly** [reply]
Making the code unsafe isn't the only reason why runtime doesn't allow heap allocation for ref struct. I think it also affects how GC traces objects. In runtime, managed pointer and object reference are not interchangeable. In the former case, the GC needs to find the head of the object being referenced before marking or moving it. CoreCLR cares very much on performance and I don't think they would ever loosen that restriction.

**Comment by MrJul** [comment]
Isn't the compiler already capable of determining if a `ref struct` crosses an await? It already knows whether it should hoist a local when compiling an iterator or async method to a state machine, and thus can error if hoisting a `ref struct` is needed.

Within iterators, it already works as expected:
```cs
// Compiles
IEnumerable<int> M1() {
    yield return 1;
    Span<int> span = stackalloc int[1];
    span[0] = 0;
    yield return 2;
}

// Does not compile
IEnumerable<int> M2() {
    yield return 1;
    Span<int> span = stackalloc int[1];
    yield return 2;
    span[0] = 0; // CS4013: instance of type 'Span<int>' cannot be used inside a nested function, query expression, iterator block or async method
}
```

In async methods, the error `CS4012: Parameters or locals of type 'Span<int>' cannot be declared in async methods or lambda expressions` is reported instead directly on the local declaration, instead of having CS4013 reported on every usage crossing an `await` statement.

I made a small test and simply removed the extra check for async methods: see https://github.com/MrJul/roslyn/commit/2528a9cd3c8aa82341e84ed6631217e3c7cfd176 After this change, they behave like iterators, reporting CS4013 only on invalid usages (the state machine code is common to iterators and async methods so that's expected).

As far as I can tell, this should be enough, other local special types affected by the check (`TypedReference`/`ArgIterator`/`RuntimeArgumentHandle`) are still handled by CS4013, and even have an [extra CS4012](https://github.com/dotnet/roslyn/blob/7dd737db504ae855dac0de45bcd11a9cf1e9e42a/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs#L74) when used as parameters. Was this check a redundant defense or is there something special with async methods that don't apply to iterators regarding ref structs?

**Reply by HaloFour** [reply]
It might be more a question of trying to describe that behavior in the spec.  It's a lot easier to forbid something outright than to try to carve out the different scenarios where something is or isn't allowed.

**Reply by bernd5** [reply]
It would be really great if we could remove that restriction.
Otherwise we are forced to write ugly workarounds...

**Comment by airbreather** [comment]
Can this proposal include `ref` locals too, please?  From my understanding, a `ref` local is **essentially** the same thing as a local variable that points to an instance of a `ref struct` (though I suppose `MemoryMarshal.CreateSpan(ref WhateverGaveMeTheRef(), 1)` isn't bad).

The only other proposal that I can find on that topic is #1173, but I don't want to support that other proposal: the author seems to insist in the comments that it's important to keep a restriction that you may not `await` or `yield return` at a point where there are `ref` local variables **in scope**, which seems rather arbitrary given that you can already `yield return` at a point where there are local variables of a `ref struct` type in scope...

**Comment by RikkiGibson** [comment]
I think there is a general opportunity for improvement here. IIRC, there are a bunch of bugs related to behavior of refs being poorly defined in async methods. Ref parameters are generally not allowed in async methods, but it's easy to introduce one for `this`. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQAwAIEEYCsBuAsAKAIQGYUBnAFwCcBXAYwpQGUUBvAlTtUhADjQBsKALIAKAJRsOXGWKoBTAGYoKACyhlx2FAHodKAORqNBlBpQRDCxaYAOEKhDDyK8qilVv50zgF8fKAEIACwiotbM5JKsKP74vkA===).

```cs
using System.Threading.Tasks;

public struct S {
    public async Task M() {
        M(ref this); // 'this' is a 'ref' parameter here
    }
    
    void M(ref S s) { }
}
```

Maybe we could simply come up with a rule that says refs are allowed but they can't be spilled and see what the implications are.

**Reply by FaustVX** [reply]
Oh, I didn't know we can pass `ref this` of a struct.
That's nice.

**Comment by bernd5** [comment]
As a workaround I wrote an unsafe struct which holds a pointer to the data.
But of course it works only for stack-locals (stack vars which do not cross await boundaries) and e.g. static fields.

```csharp
unsafe struct UnsafeRef<T>
{
    void* ptr;        
    
    public UnsafeRef(ref T t)
    {
        ptr = Unsafe.AsPointer(ref t);
    }

    public UnsafeRef(in T t, byte dummy = 0) : this(ref Unsafe.AsRef(t))
    {
    }
    
    public ref T Value
    {
        get{
            return ref Unsafe.AsRef<T>(ptr);
        }
    }
}
```

To create such a ref var you can write: `new UnsafeRef<int>(ref xxx)` and to use it: `ref myRefVar.Value`

**Reply by CodingMadness** [reply]
For what is the "byte dummy=0" variable?

**Reply by RikkiGibson** [reply]
It lets you fake overloading a constructor on in/ref, which isn't normally possible.
